test/e2e/podtimingcontroller/podtimingcontroller.go

```diff
@@ -0,0 +1,84 @@
+// +build e2e
+
+package podtimingcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+func SetupWithManager(ctx context.Context, mgr ctrl.Manager, log logr.Logger, artifactDir string) error {
+	r := &podTimingReconciler{
+		observedPhases: map[string][]Phase{},
+		client:         mgr.GetClient(),
+	}
+	go func() {
+		<-ctx.Done()
+		r.observedPhasesLock.Lock()
+		serialized, err := json.Marshal(r.observedPhases)
+		if err != nil {
+			log.Error(err, "failed to serialize pod phases")
+			return
+		}
+		if err := ioutil.WriteFile(path.Join(artifactDir, "pod_phases.json"), serialized, 0644); err != nil {
```

Comment:
Is there a tool that can visualize this output?

Reply:
No, the idea for this change was to just start off with something that collects the data (and hear what you think about the approach). We can then build something that e.G. creates a HTML timeline out of that (or even embedd it into the artifacts like the OCP intervals like e.G. here: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.9-e2e-gcp-fips/1439715993600397312)